### PR TITLE
feat(form-workspaces-be/6): add move forms from a single source workspace functionality

### PR DESCRIPTION
**Pull request by hanstirtaputra**
Builds on #4254 

Part of #4039 

## Solution
Add endpoint logic for POST `api/v3/admin/workspaces/{workspaceId}/forms/move` for moving forms from one source workspace to one destination workspace

**Breaking Changes** 
- [x] No - this PR is backwards compatible

### CODE REVIEW

File: src/app/models/workspace.server.model.ts [hunk 1]
```diff
@@ -86,6 +91,38 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
+    formIds,
+    session,
+  }: {
+    workspaceId: IWorkspaceSchema['_id']
+    formIds: IFormSchema['_id'][]
+    session?: ClientSession
+  }) {
+    await this.updateOne(
+      { _id: workspaceId },
+      { $pullAll: { formIds } },
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts:105]: cool!

File: src/app/models/workspace.server.model.ts [hunk 2]
```diff
@@ -86,6 +91,38 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
+    formIds,
+    session,
+  }: {
+    workspaceId: IWorkspaceSchema['_id']
+    formIds: IFormSchema['_id'][]
+    session?: ClientSession
+  }) {
+    await this.updateOne(
+      { _id: workspaceId },
+      { $addToSet: { formIds: { $each: formIds } } },
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts:121]: also cool!

File: src/app/modules/workspace/__tests__/workspace.controller.spec.ts [hunk 3]
```diff
@@ -328,4 +328,120 @@ describe('workspace.controller', () => {
...
+      await WorkspaceController.updateFormsWorkspace(
+        MOCK_REQ,
+        mockRes,
+        jest.fn(),
+      )
+
+      expect(mockRes.json).toHaveBeenCalledWith(MOCK_WORKSPACE)
+    })
+
+    it('should return 403 when workspace is not found', async () => {
```
Comment (mantariksh) [inline src/app/modules/workspace/__tests__/workspace.controller.spec.ts:371]: looks like wrong test description?

File: src/app/modules/workspace/__tests__/workspace.controller.spec.ts [hunk 4]
```diff
@@ -328,4 +328,120 @@ describe('workspace.controller', () => {
...
+        MOCK_REQ,
+        mockRes,
+        jest.fn(),
+      )
+
+      expect(mockRes.status).toBeCalledWith(403)
+      expect(mockRes.json).toBeCalledWith({ message: mockErrorString })
+    })
+
+    it('should return 404 when workspace is not found', async () => {
```
Comment (mantariksh) [inline src/app/modules/workspace/__tests__/workspace.controller.spec.ts:390]: probably need to swap this with prev test

File: src/app/modules/workspace/workspace.controller.ts [hunk 5]
```diff
@@ -247,22 +247,57 @@ export const deleteForms: ControllerHandler<
...
  * @returns 404 when the workspace does not exist or belong to the user
  * @returns 422 when user of given id cannnot be found in the database
  * @returns 500 when database errors occur
  */
-export const moveForms: ControllerHandler<
-  { sourceWorkspaceId: string },
-  unknown,
-  any | ErrorDto,
-  { formIds: any[]; destWorkspaceId: string }
+export const updateFormsWorkspace: ControllerHandler<
```
Comment (mantariksh) [inline src/app/modules/workspace/workspace.controller.ts:255]: not a very clear function name, maybe `moveFormsToWorkspace`?

File: src/app/models/workspace.server.model.ts [hunk 6]
```diff
@@ -86,6 +91,38 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
   }
 
+  WorkspaceSchema.statics.removeFormIdsFromWorkspace = async function ({
+    workspaceId,
+    formIds,
+    session,
+  }: {
+    workspaceId: IWorkspaceSchema['_id']
+    formIds: IFormSchema['_id'][]
+    session?: ClientSession
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts:101]: if we always expect this operation to take place within a transaction, we should make this param mandatory

File: src/app/modules/workspace/workspace.controller.ts [hunk 7]
```diff
@@ -247,22 +247,57 @@ export const deleteForms: ControllerHandler<
...
+      workspace
+        ? res.status(StatusCodes.OK).json(workspace)
+        : res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({
+            message:
+              'Sorry something went wrong, we are unable to move the forms to another worksapce',
+          }),
     )
+    .mapErr((error) => {
+      logger.error({
+        message: 'Error moving forms to anotehr workspace',
```
Comment (mantariksh) [inline src/app/modules/workspace/workspace.controller.ts:289]:
```suggestion
        message: 'Error moving forms to another workspace',
```

File: src/app/models/workspace.server.model.ts [hunk 8]
```diff
@@ -86,6 +91,38 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
+  }) {
+    await this.updateOne(
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts:103]: why don't we use `findOneAndUpdate` so we can return the updated document to the caller?

File: src/app/models/workspace.server.model.ts [hunk 9]
```diff
@@ -86,6 +91,38 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
+  WorkspaceSchema.statics.addFormIdsToWorkspace = async function ({
+    workspaceId,
+    formIds,
+    session,
+  }: {
+    workspaceId: IWorkspaceSchema['_id']
+    formIds: IFormSchema['_id'][]
+    session?: ClientSession
+  }) {
+    await this.updateOne(
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts:119]: we can use `findOneAndUpdate`? then we don't have to make another db query to retrieve the updated destination workspace at the end

File: src/app/modules/workspace/workspace.controller.ts [hunk 10]
```diff
@@ -247,22 +247,57 @@ export const deleteForms: ControllerHandler<
...
+      }),
+    )
+    .map((workspace) =>
+      workspace
+        ? res.status(StatusCodes.OK).json(workspace)
+        : res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({
+            message:
+              'Sorry something went wrong, we are unable to move the forms to another worksapce',
+          }),
     )
```
Comment (mantariksh) [inline src/app/modules/workspace/workspace.controller.ts]: hmm, this is messy. at this point we should be sure that the operation has succeeded, all errors should flow to the `mapErr` block below. can we get `WorkspaceService.moveForms` to return an error if the workspace can't be found? using `findOneAndUpdate` in the model layer will probably help

**Comment by mantariksh** [comment]
how do we prevent form admins from submitting IDs of forms they don't own, which would result in their workspaces containing forms they shouldn't be able to access?

**Comment by hanstirtaputra** [comment]
Closing this PR as it is a special case of `updateMany` in #4448 . Will just keep #4448 instead.

**Pull request closed without merging**